### ISSUE TITLE
Postgres, Redshift: POSIX Pattern Matching Expressions are not Parsed

### ISSUE THREAD

**Issue by PLBMR** [sqlfluff/sqlfluff#2731]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Followup from #2722 .

While [Postgres](https://www.postgresql.org/docs/14/functions-matching.html#FUNCTIONS-POSIX-REGEXP) and [Redshift](https://docs.aws.amazon.com/redshift/latest/dg/pattern-matching-conditions-posix.html) both support POSIX Pattern Matching Expressions, they are not parsed in our current dialects. Say `example.sql` is set to 

```sql
WITH cleaned_bear_financial_branch AS (
    SELECT
        branch_id,
        -- ~ is POSIX operator
        TO_NUMBER(CASE WHEN honey_numerical_code ~ '[0-9]{0,7}.?[0-9]{0,2}' THEN honey_numerical_code ELSE NULL END, '24601') AS honey_numerical_code
    FROM bear_financial_branch
)

SELECT branch_id
FROM cleaned_bear_financial_branch
LIMIT 10;
```

We get the following parse errors when `lint`ing in `postgres` and `redshift`:

```bash
>>> sqlfluff lint --dialect=postgres example.sql 
== [example.sql] FAIL                                                                                                                                                             
L:   5 | P:  19 |  PRS | Line 5, Position 19: Found unparsable section: "CASE WHEN honey_numerical_code ~ '[0-9]{..."
All Finished 📜 🎉!
>>> sqlfluff lint --dialect=redshift example.sql
== [example.sql] FAIL                                                                                                                                                             
L:   5 | P:  19 |  PRS | Line 5, Position 19: Found unparsable section: "CASE WHEN honey_numerical_code ~ '[0-9]{..."
All Finished 📜 🎉!
```

### Expected Behaviour

POSIX pattern matching expressions are parsed when `lint`ing in the `postgres` and `redshift` dialects.

### Observed Behaviour

POSIX pattern matching expressions are not parsed when `lint`ing in the `postgres` and `redshift` dialects.

### How to reproduce

1. Install requirements: `pip install -r dbt==0.20.0 sqlfluff==0.10.1 markupsafe==2.0.1`
2. Create `example.sql` in your working directory with the content specified above.
3. Run `sqlfluff lint --dialect=postgres example.sql`. Confirm that you get a parsing error.
4. Run `sqlfluff lint --dialect=redshift example.sql`. Confirm that you get a parsing error.

### Dialect

`postgres` and `redshift` dialects.

### Version

`sqlfluff==0.10.1`.

### Configuration

`Python 3.8.2`

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by PLBMR** [comment]
My recommended dev plan:
1. Create `PosixExpressionSegment`, which can be added to the `match_grammar` of `PatternMatchExpression` in `src/sqlfluff/dialects/dialect_postgres.py`.
2. Implement `PosixExpressionSegment` which is a `PosixOperatorSegment` (`~`, `!~`, `~*`, `!~*`) followed by a Regular Expression in a string literal.
3. Replace `PosixOperatorSegment` in the `redshift` dialect to only support (`~`, `!~`). This is because `redshift` [does not currently support case-insensitive POSIX operations](https://docs.aws.amazon.com/redshift/latest/dg/pattern-matching-conditions-posix.html).

**Comment by tunetheweb** [comment]
@PLBMR I resume this can be closed thanks to #2732?
